frontend/public/components/graphs/query-browser.tsx

```diff
@@ -0,0 +1,234 @@
+      const differingLabelKeys = _.filter(allLabelKeys, k => _.uniqBy(allLabels, k).length > 1);
+
+      const newGraphData = [];
+      _.each(result, d => {
+        const labels = _.omit(d.metric, '__name__');
+
+        // If metric prop is specified, ignore all other metrics
+        if (metric && _.some(labels, (v, k) => _.get(metric, k) !== v)) {
+          return;
+        }
+
+        // Just show labels that differ between metrics to keep the name shorter
+        const name = _.map(_.pick(labels, differingLabelKeys), (v, k) => `${k}="${v}"`).join(',');
+
+        const values = _.map(d.values, v => ({
+          x: new Date(v[0] * 1000),
+          y: parseFloat(v[1]),
+        }));
+
+        // The data may have missing values, so we fill those gaps with nulls so that the graph correctly shows the
+        // missing values as gaps in the line
+        const start = Number(_.get(values, '[0].x'));
+        const end = Number(_.get(_.last(values), 'x'));
+        const step = span / samples;
+        _.range(start, end, step).map((t, i) => {
+          const x = new Date(t);
+          if (_.get(values, [i, 'x']) > x) {
+            values.splice(i, 0, {x, y: null});
+          }
+        });
+
+        newGraphData.push({name, values});
+      });
+
+      setGraphData(newGraphData);
+
+      if (onDataUpdate) {
+        onDataUpdate(result);
+      }
+    }
+    setUpdating(false);
+
+    // Only trigger when data is updated
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [data]);
+
+  const onSpanChange = newSpan => {
+    setDomain(undefined);
+    setSpan(newSpan);
+    setUpdating(true);
+  };
+
+  const onZoom = ({x, y}) => {
+    setDomain({x, y});
+    setSpan(x[1] - x[0]);
+    setUpdating(true);
+  };
+
+  const graphDomain = domain || {x: [Date.now() - span, Date.now()], y: undefined};
+
+  return <div className="query-browser__wrapper">
+    <div className="query-browser__header">
+      <SpanControls defaultSpanText={defaultSpanText} onChange={onSpanChange} span={span} />
+      <div className="query-browser__loading">
+        {updating && <LoadingInline />}
+      </div>
+      <div className="query-browser__external-link">
+        {GraphLink}
+      </div>
+    </div>
+    {error && <div className="alert alert-danger query-browser__error">
+      <span className="pficon pficon-error-circle-o" aria-hidden="true"></span>{error.message}
+    </div>}
+    <Graph colors={colors} data={graphData} domain={graphDomain} onZoom={onZoom} query={query} />
+  </div>;
+};
+
+type Domain = {
+  x: [number, number];
+  y: [number, number];
+};
+
+type GraphDataPoint = {
+  name?: string;
+  x: Date;
+  y: number;
+};
+
+type GraphDataMetric = {
+  metric: {[key: string]: string}[];
+  values: GraphDataPoint[];
+};
+
+type GraphProps = {
+  colors: string[];
+  data: {
+    name: string;
+    values: GraphDataPoint[];
+  }[];
+  domain: Domain;
+  onZoom: (Domain) => void;
```

Comment:
I believe this is naming the parameter `Domain`, not specifying the type.

Reply:
Oops. Fixed by #1614